### PR TITLE
EC2: Added growpart script

### PR DESCRIPTION
**Pull request by jwendell**
This is an attempt to fix issue #237

The script "growpart" was added as an asset, because
Debian's Jessie version (inside cloud-utils) package is
outdated and buggy.

Then "expand-root" init script was modified to call growpart
before calling resize2fs. In fact, calling resize2fs without
resizing the partition first is useless.

The task was named 'AddWorkaroundGrowpart' because it must
go away in the future in favor of using recent stuff in cloud-utils.

Currently even the official images suffer this issue, that makes
the system use only 8 GiB, even if the user chooses a bigger
storage size inside AWS.

### CODE REVIEW

**Comment by jwendell** [comment]
Forgot to say: I'm *not* using cloud-init plugin.
If someone is using it, probably it's broken too, because there's no growpart script.

File: bootstrapvz/common/assets/init.d/expand-root [hunk 1]
```diff
@@ -13,8 +13,19 @@
...
 
-device_path="/dev/xvda"
+growpart=$(which growpart) || {
+	$logger "growpart was not found on PATH. Unable to expand size."
+	exit 0
+}
 
+root_index="0"
+
+$growpart /dev/xvda $root_index || {
```
Comment (andsens) [inline bootstrapvz/common/assets/init.d/expand-root]: Wouldn't just calling `growpart` be fine here? We just established that it is on `$PATH`.

File: bootstrapvz/common/assets/init.d/expand-root [hunk 2]
```diff
@@ -13,8 +13,19 @@
 prog=$(basename $0)
 logger="logger -t $prog"
 
-device_path="/dev/xvda"
+growpart=$(which growpart) || {
+	$logger "growpart was not found on PATH. Unable to expand size."
+	exit 0
```
Comment (andsens) [inline bootstrapvz/common/assets/init.d/expand-root]: Shouldn't we exit with something other than `0`?

File: bootstrapvz/common/assets/init.d/expand-root [hunk 3]
```diff
@@ -13,8 +13,19 @@
 prog=$(basename $0)
 logger="logger -t $prog"
 
-device_path="/dev/xvda"
+growpart=$(which growpart) || {
```
Comment (andsens) [inline bootstrapvz/common/assets/init.d/expand-root]: We shouldn't use `which`. [Consider using `type` or `hash` instead](http://stackoverflow.com/a/677212/339505).

File: bootstrapvz/common/assets/init.d/expand-root [hunk 4]
```diff
@@ -13,8 +13,19 @@
...
+}
 
+root_index="0"
+
+$growpart /dev/xvda $root_index || {
+	$logger "growpart failed. Unable to expand size."
+	exit 0
+}
+
+device_path="/dev/xvda$root_index"
```
Comment (andsens) [inline bootstrapvz/common/assets/init.d/expand-root]: I think it's better to allow changing the entire device path. This way we can easily switch to `/dev/by-uuid/` or something else if we want to.

File: bootstrapvz/providers/ec2/__init__.py [hunk 5]
```diff
@@ -67,6 +67,7 @@ def resolve_tasks(taskset, manifest):
 	                boot.DisableGetTTYs,
 	                tasks.network.EnableDHCPCDDNS,
 	                initd.AddExpandRoot,
+	                tasks.packages.AddWorkaroundGrowpart,
```
Comment (andsens) [inline bootstrapvz/providers/ec2/__init__.py]: Shouldn't the workaround task only be present when bootstrapping `>=jessie`?

**Comment by andsens** [comment]
Sorry it took so long for me to answer this. July has been quite busy for me.
Could you check the comments I added?

File: bootstrapvz/common/assets/init.d/expand-root (hunk 2 above)
Comment (jszwedko) [inline bootstrapvz/common/assets/init.d/expand-root]: Would exiting `1` work? Or did you have another code in mind.

File: bootstrapvz/common/assets/init.d/expand-root (hunk 3 above)
Comment (jszwedko) [inline bootstrapvz/common/assets/init.d/expand-root]: Agreed :+1:

File: bootstrapvz/common/assets/init.d/expand-root (hunk 1 above)
Comment (jszwedko) [inline bootstrapvz/common/assets/init.d/expand-root]: Also agreed :+1:

**Comment by jszwedko** [comment]
Bump @jwendell @andsens I can pick up this work if needed -- we've been frustrated by this issue as well.

File: bootstrapvz/common/assets/init.d/expand-root (hunk 3 above)
Comment (andsens) [inline bootstrapvz/common/assets/init.d/expand-root]: In my other project (homeshick) I have actually gathered [a list of exit codes](https://github.com/andsens/homeshick/blob/e8d021c10c7a54fba0a81a7995ca0148c07f3a9b/lib/exit_status.sh) from different sources. It seems `1` is the only one applicable.

File: bootstrapvz/common/assets/init.d/expand-root (hunk 2 above)
Comment (andsens) [inline bootstrapvz/common/assets/init.d/expand-root]: In my other project (homeshick) I have actually gathered [a list of exit codes](https://github.com/andsens/homeshick/blob/e8d021c10c7a54fba0a81a7995ca0148c07f3a9b/lib/exit_status.sh) from different sources. It seems `1` is the only one applicable.

**Comment by andsens** [comment]
Sure @jszwedko, that'd be fine by me.

File: bootstrapvz/common/assets/init.d/expand-root (hunk 2 above)
Comment (jszwedko) [inline bootstrapvz/common/assets/init.d/expand-root]: :+1:

File: bootstrapvz/providers/ec2/__init__.py (hunk 5 above)
Comment (jszwedko) [inline bootstrapvz/providers/ec2/__init__.py]: Yeah, I think this is only applicable to `jessie`

**Comment by jwendell** [comment]
Sure, go ahead :)

File: bootstrapvz/common/assets/init.d/expand-root (hunk 4 above)
Comment (jszwedko) [inline bootstrapvz/common/assets/init.d/expand-root]: `root_index` is set above here -- did you envision `device_path` as a whole being configurable outside of this init script? Or just moving `/dev/xvda` to another variable here?

Comment (jszwedko) [inline bootstrapvz/common/assets/init.d/expand-root]: Ignore this, I see the `sed`ing happening below.

Comment (jszwedko) [inline bootstrapvz/common/assets/init.d/expand-root]: Is it possible to determine the disk and partition given the device path? I can `sed` just the device path into here, but `growpart` requires the disk and partition as arguments.

Comment (andsens) [inline bootstrapvz/common/assets/init.d/expand-root]: Just make it two variables :-)

**Comment by andsens** [comment]
Closing, in favor of #269. Thank you for laying the foundation @jwendell! :+1:

**Pull request closed without merging**